### ISSUE TITLE
ROCm 3.5 link error/ROCm hipStreamAddCallback crash

### ISSUE THREAD

**Issue by tomdeakin** [illuhad/hipSYCL#265]
Have you seen this error before?
```
[ 50%] Linking CXX shared library libhipSYCL_rocm.so
/opt/rh/devtoolset-8/root/usr/lib/gcc/x86_64-redhat-linux/8/../../../../bin/ld: cannot find -lmcwamp
/opt/rh/devtoolset-8/root/usr/lib/gcc/x86_64-redhat-linux/8/../../../../bin/ld: cannot find -lhc_am
clang-9: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/sycl/libhipSYCL_rocm.so] Error 1
make[1]: *** [src/sycl/CMakeFiles/hipSYCL_rocm.dir/all] Error 2
make: *** [all] Error 2
```

I'm not inside the `devtoolset-8`, so there might be some hard coded paths curtesy of the ROCm install via the `yum` repository.

**Comment by illuhad** [comment]
No.. But AMD's ROCm packages for various distributions can vary quite a bit. Linking against these libraries used to be necessary, so this is done by default by `syclcc`. I suspect that these libraries are either not there entirely with your ROCm distribution, or their content is statically linked into some other library.
Try removing these libraries from the `ROCM_LINK_LINE` setting inside `$HIPSYCL_INSTALL_PREFIX/etc/hipSYCL/syclcc.json`. This is from where `syclcc` draws its default linker flags.

**Comment by tomdeakin** [comment]
Thanks - I had to remove it directly in `CMakeList.txt` but it managed to build hipSYCL.

I find a different error when building SYCL code:
```
error: unable to load plugin '/nfs/software/x86_64/hipsycl/master-d4064619/bin/../lib/libhipSYCL_clang.so': '/nfs/software/x86_64/hipsycl/master-d4064619/bin/../lib/libhipSYCL_clang.so: undefined symbol: _ZNK4llvm10ModulePass17createPrinterPassERNS_11raw_ostreamERKSs'
```

**Comment by illuhad** [comment]
This happens when there is an ABI mismatch between the clang plugin and the clang you are loading it into. E.g., clang plugin is compiled against a different clang version, or cmake has not correctly detected settings that LLVM was built with (e.g. `-fno-rtti`, `-fno-exceptions`). The latter I have seen with some AMD provided clang distributions that don't seem to correctly advertise their build flags.

**Comment by tomdeakin** [comment]
That's interesting because I was building with my own clang 9 build. Building with GCC instead works around it.

Have you seen segmentation faults inside HIP?
The `gdb` backtrace identifies it here:
```
#0  0x00007ffff7acf30f in ihipStreamCallback(_cl_event*, int, void*) () from /opt/rocm/lib/libamdhip64.so.3
```

**Comment by illuhad** [comment]
No... this might be a regression in ROCm 3.5. I would recommend to check with a HIP test program whether `hipStreamAddCallback()` works correctly, and if not, open an issue with AMD's HIP repository.

**Comment by FilipVaverka** [comment]
I have the same problem. I've tried to build/run tests "HIP/tests/src/runtimeApi/stream/*.cpp" against official ROCm 3.5.1 packages. The tests seem to pass fine. I should note that I had to compile those manually using "hipcc" as I was unable to get their "HIT" build/tests system to work (seems like it needs static build of ROCCLR libaries).

**Comment by illuhad** [comment]
I can reproduce the `hipStreamAddCallback` issue with our experimental ROCm 3.5.x packages. No idea what is causing that..

**Comment by tomdeakin** [comment]
I opened [an issue](https://github.com/ROCm-Developer-Tools/HIP/issues/2119#issuecomment-651683766) for it against HIP directly.

**Comment by illuhad** [comment]
@tomdeakin Thanks, I have subscribed to that issue :)

**Comment by FilipVaverka** [comment]
I did bit more digging on this `hipStreamAddCallback` issue. It seems it gets called with `null` stream. In fact whole `task_graph_node` is created with stream for which `_stream->get_stream()` returns 0. This propagates through `task_graph_node::submit()` and `hipStreamAddCallback` down to `ihipStreamCallback` callback inside of HIP, where invalid `amd::ScopedLock` is constructed from it and finally crash occurs.

**Comment by illuhad** [comment]
@FilipVaverka  Thank you, that is valuable information! I will investigate.

**Comment by illuhad** [comment]
@FilipVaverka Okay, I think we are on the wrong track here. I put in a `printf()` to dump the addresses contained by the `stream_ptr` objects before the only `hipStreamAddCallback` in `task_graph.cpp`:
```
./test
0x523110
0x523110
0x56a3f0
Speicherzugriffsfehler (Speicherabzug geschrieben)
```
The segmentation fault then shows our familiar backtrace. As you can see, none of the streams are `nullptr`. I think the line
```
hipStreamAddCallback (stream=0x0,
callback=0x7ffff177eae0 <hipsycl::sycl::detail::task_done_callback(ihipStream_t*, hipError_t, void*)>, userData=0x566de0, flags=0)
```
from your backtrace instead shows that hipSYCL has submitted a callback to the default stream. In CUDA/HIP, the first argument of `hipStreamAddCallback` is `hipStream_t` which is a typedef for int. A value of 0 (integer 0, not NULL ;) ). Denotes the default stream.

However, realizing that the AMD's HIP test cases for `hipStreamAddCallback` don't seem to test the default stream, I think I can reproduce the issue by modifying the test case slightly:

```c++
#include <stdio.h>
#include <thread>
#include <chrono>
#include "hip/hip_runtime.h"


#ifdef __HIP_PLATFORM_HCC__
#define HIPRT_CB
#endif

constexpr size_t N = 1024*1024;

__global__ void vector_square(float* C_d, float* A_d, size_t N) {
    size_t offset = (blockIdx.x * blockDim.x + threadIdx.x);
    size_t stride = blockDim.x * gridDim.x;

    for (size_t i = offset; i < N; i += stride) {
        C_d[i] = A_d[i] * A_d[i];
    }
}

// To make this file independent of AMD test framework
void HIPCHECK(hipError_t err){
    if(err != hipSuccess)
        printf("Error!\n");
}


float *A_h, *C_h;
bool cbDone = false;

void Callback(hipStream_t stream, hipError_t status, void* userData) {
    for (size_t i = 0; i < N; i++) {
        if (C_h[i] != A_h[i] * A_h[i]) {
            printf("Warning: Data mismatch %zu", i);
        }
    }
    printf("PASSED!\n");
    cbDone = true;
}

int main(int argc, char* argv[]) {
    float *A_d, *C_d;
    size_t Nbytes = N * sizeof(float);

    A_h = (float*)malloc(Nbytes);
    HIPCHECK(A_h == 0 ? hipErrorOutOfMemory : hipSuccess);
    C_h = (float*)malloc(Nbytes);
    HIPCHECK(C_h == 0 ? hipErrorOutOfMemory : hipSuccess);

    // Fill with Phi + i
    for (size_t i = 0; i < N; i++) {
        A_h[i] = 1.618f + i;
    }

    HIPCHECK(hipMalloc(&A_d, Nbytes));
    HIPCHECK(hipMalloc(&C_d, Nbytes));

    hipStream_t mystream = 0; //<- Changed this and removed the next line
    //HIPCHECK(hipStreamCreateWithFlags(&mystream, hipStreamNonBlocking));

    HIPCHECK(hipMemcpyAsync(A_d, A_h, Nbytes, hipMemcpyHostToDevice, mystream));

    const unsigned blocks = 512;
    const unsigned threadsPerBlock = 256;
    hipLaunchKernelGGL((vector_square), dim3(blocks), dim3(threadsPerBlock), 0, mystream, C_d, A_d,
                       N);

    HIPCHECK(hipMemcpyAsync(C_h, C_d, Nbytes, hipMemcpyDeviceToHost, mystream));
    HIPCHECK(hipStreamAddCallback(mystream, Callback, NULL, 0));
    
    while (!cbDone) std::this_thread::sleep_for(std::chrono::milliseconds(10));
}

```

**Comment by FilipVaverka** [comment]
That might be it. Looking into ROCCLR code (`HIP/rocclr/hip_stream.cpp`), I can't find anything that would look like translation from HIP stream handle to actual object pointer. Its just `hipStreamAddCallback` which packs handle into `StreamCallback` object and in `ihipStreamCallback` its unpacked and casted to `hip::Stream *`. This obviously fails if stream handle is 0.

**Comment by illuhad** [comment]
It used to work fine in earlier ROCm versions. At least it didn't crash. Either they changed the implementation then or the callback didn't work previously (which could go unnoticed if there are enough other tasks on non-default streams where the callbacks work) and just didn't crash...

**Comment by illuhad** [comment]
Issue is known and will be fixed in ROCm 3.6: https://github.com/ROCm-Developer-Tools/HIP/issues/2119#issuecomment-652556974

I think we can close now here, because it's not an hipSYCL issue and all that's left to be done is to install a new ROCm. Please reopen if there is anything we need to do from the hipSYCL side.
